uiowa/uiowa
# Parent title field

Resolves #2048 

Compound title field to share across blocks. Includes the title itself, a way to control its heading size, and uses this to set child heading sizes (accordion headings, article titles, etc.) for proper DOM hierarchy. If no title is needed/wanted, the heading size can be set for the children directly.

## Testing

Pull
`composer install`
`blt frontend`
sync a site
Test on existing or new blocks:
* Articles
* Events
* Text Area
* Image Gallery
* Collection
* Slider
Check the following:
* Include a title and set different heading levels
  * For blocks with child elements, they should appear at the appropriate level below the title
* Try visually hiding the title, with various heading levels
  * The title should be hidden
  * The title should appear in the DOM on inspection
  * For blocks with child elements, they should appear at the appropriate level below the title
* Try setting the "child heading level" when the title is not filled
  * No title should appear on render, either visually nor in the DOM
  * The child elements' heading levels should render at the appropriate level

<details>
<summary>
Old notes
</summary>

Roadmap for parent title field.
* What is needed, assuming it requires a compound field?
  - [x] Title (text)
    * Plain, allow for formatting for future options (highlighting)?
      * For future options, do we need a minimal HTML field or similar, or can we use a plain, long field with tags like `<hl>` (or whatever is in place) that would be replaced with complete markup?
  - [x] Size (select list)
    * One field. Logic applies it to title if it is used, or children if not.
        * Any use case where you would _want_ to set both title and children sizes separately, that wouldn't break hierarchy?
  - [x] Display title (boolean)
    * How should this interact with children? Have a hidden title heading over visible children headings, or does this essentially mean don't use the title field?
    * In addition to possible formatting in the text field that might apply to only part of the title, this would apply to all.
    * This can be added later more easily than others, as it would be adding a field or view modes, but the overall field won't be affected.
* What logic needs to be created
  - [x] Display/no display title
    * How does the hide title boolean (above) need to function?
  - [x] Control heading size of children
    * This is in place for collections/sliders currently (which both use paragraphs), and needs to be expanded for other block types.
  - [x] Update hook for blocks using existing Heading size field
    * Would need to set this new field for any block that uses this old field, and set heading size appropriately.
      * And probably need a two-step process. 1) update config to add new heading field alongside old heading size, and update hook to appropriately set the new field, and 2) Remove the old field
    * For any blocks which don't currently have this field, need to set this field (or make sure any NULLs are appropriately handled).
      * Collections (field was hidden--default to h2)
      * Text Area
      * Image Gallery
* What blocks need to use it?
  - [x] People/articles/events
    * Views-type collections, for a title above the view and setting the child heading sizes appropriately.
    * Accordions (and future tab potential) set a title above and child headings. Descriptive lists the title and its heading size only.
    * Title above, and slider heading sizes.
  - [x] Image gallery
  - [x] Text block(?)
    * Title above. Possibly unset heading sizes on that level or above?
        * e.g. Text block with an `<h2>` as a title has access only to 'Heading 3' and below from the style dropdown. Simple approach, would be possible to break by going into Source, but would provide some guidance to users (and potential confusion if they don't understand, and suddenly can't find a 'Heading 3.')

## PR current state
 * Layout Builder Custom defines a fieldtype and its formatter/widget.
 * Config sets this field on several blocktypes.
 * uids_base.theme paragraph preprocess sets heading sizes for collections and sliders.
 * sitenow_articles, sitenow_people, sitenow_events block preprocesses set heading sizes for their respective blocks.
 * Update hook sets fields appropriately for blocks which have a heading_size or existing title/heading_size field prior to those fields being removed from the block.
</details>
